### PR TITLE
Add schemas for plans

### PR DESCRIPTION
**Pull request by thrix**
Todo
- [x] context
- [x] discover
- [x] prepare
- [x] provision
- [x] execute
- [x] finish
- [x] report

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

**Comment by thrix** [comment]
@happz @lukaszachy @psss this now passes for me and validates all available plans.

One thing that is fairly unclear is the role of the core attributes, I would expect they can be part of any fmf node, i.e. a plan, test or story. But according to @lukaszachy they can be in any leaf of the node data??

File: tests/unit/test_schemas.py [hunk 1]
```diff
@@ -27,16 +27,36 @@ def stories_schema():
...
+            'finish/ansible', 'finish/shell',
+            'prepare/ansible', 'prepare/install', 'prepare/shell',
+            'provision/artemis', 'provision/connect', 'provision/container',
+            'provision/local', 'provision/virtual',
+            'report/display', 'report/html'
+            ):
         schema_file = os.path.join(SCHEMADIR, f'{schema_name}.yaml')
-        schema = YAML(typ="safe").load(
-            open(schema_file, encoding="utf-8"))
+        schema = YAML(
```
Comment (thrix) [inline tests/unit/test_schemas.py:55]: formatting

File: tmt/schemas/common.yaml [hunk 2]
```diff
@@ -148,11 +214,41 @@ definitions:
...
+
+  # helper
+  url:
+    type: string
+    pattern: "^https?://"
+
   # a url, path or fmf_id
   url_path_or_fmf_id:
     oneOf:
-      - type: string
```
Comment (thrix) [inline tmt/schemas/common.yaml:154]: hmm, what the hell is this change :)

Comment (thrix) [reply tmt/schemas/common.yaml:154]: nm, it is correct, url become a common definition.

File: tests/unit/test_schemas.py [hunk 3]
```diff
@@ -27,13 +27,30 @@ def stories_schema():
     return YAML(typ="safe").load(open(schema_file, encoding="utf-8"))
 
 
+@pytest.fixture
```
Comment (guy9050) [inline tests/unit/test_schemas.py:30]: Would it make sense to set a scope to this fixture?

Comment (happz) [reply tests/unit/test_schemas.py:30]: Probably not. Fixture scope takes care of clearing fixture when pytest leaves the given scope (module, package, etc.), and this particular fixture is not harmless, does not spoil other tests or fixtures, and it's not in conflict with any other fixture.

Comment (thrix) [reply tests/unit/test_schemas.py:30]: thanks Milos.

**Comment by thrix** [comment]
@psss @lukaszachy @happz this one needs finally review to unblock #1178

I had one discussion topic, but I moved it to #1182 discussion topics, and I guess we will discuss it briefly on the hacking session.

**Comment by lukaszachy** [comment]
@thrix Please add 'id' to the schema (it was added to the core attributes just recently)

File: tmt/schemas/provision/hardware.yml [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+properties:
+  anyOf:
+    - "$ref": "#/definitions/hw_boot"
+    - "$ref": "#/definitions/hw_cpu"
+    - "$ref": "#/definitions/hw_disks"
+    - "$ref": "#/definitions/hw_networks"
+    - "$ref": "#/definitions/hw_virtualization"
+    - "$ref": "#/definitions/hw_block"
+    - "$ref": "#/definitions/hw_and"
+    - "$ref": "#/definitions/hw_or"
```
Comment (psss) [inline tmt/schemas/provision/hardware.yml:218]: The `memory` key is missing here. Tested with a simple plan example:

```yaml
provision:
    how: artemis
    hardware:
        memory: 8GB
```

Aren't the key names missing here? Shouldn't the definition here be similar as in `hw_block` above?

File: tmt/schemas/provision/virtual.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  image:
+    type: string
+
+  memory:
+    type: integer
+
+  user:
+    type: string
+
```
Comment (psss) [inline tmt/schemas/provision/virtual.yaml:33]: Missing keys `disk`, `connection` and `arch`.

File: tmt/schemas/report/html.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+type: object
+additionalProperties: false
+
+properties:
+
+  how:
+    type: string
+    enum:
+      - html
```
Comment (psss) [inline tmt/schemas/report/html.yaml:20]: Missing key `open`.

**Review by psss — DISMISSED** [review]
Thanks much for preparing this. This must have been a lot of work! Looks good in general. Added a bunch of comments and questions.

File: tests/unit/test_schemas.py [hunk 7]
```diff
@@ -26,13 +26,30 @@ def stories_schema():
...
     store = {}
 
-    for schema_name in ('common', 'core'):
+    for schema_name in (
+            'common', 'core',
+            'discover/fmf', 'discover/shell',
+            'execute/tmt', 'execute/upgrade',
+            'finish/ansible', 'finish/shell',
+            'prepare/ansible', 'prepare/install', 'prepare/shell',
+            'provision/artemis', 'provision/connect', 'provision/container',
```
Comment (psss) [inline tests/unit/test_schemas.py:49]: Missing `provision/hardware`. The file should be renamed from `hardware.yml` to `hardware.yaml`.

File: tmt/schemas/provision/hardware.yml (hunk 4 above)
Comment (happz) [reply tmt/schemas/provision/hardware.yml:218]: I'm not sure, it looks correct, but I am too seeing validation errors with your example. I don't know why, yet, but the schema should work, it's been used by Artemis for months. Maybe there's a typo or a copy/paste error.

**Comment by thrix** [comment]
> @thrix Please add 'id' to the schema (it was added to the core attributes just recently)

@lukaszachy thanks, added

File: tmt/schemas/common.yaml [hunk 8]
```diff
@@ -8,6 +8,54 @@ $id: /schemas/common
...
+        type: string
+
+      name:
+        type: string
+
+      order:
+        $ref: "/schemas/core#/definitions/order"
+
+      playbook:
+        $ref: "/schemas/common#/definitions/array_of_strings"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/common.yaml:33]: thanks!

File: tmt/schemas/common.yaml [hunk 9]
```diff
@@ -148,11 +214,41 @@ definitions:
         items:
           type: string
 
+  # common definition of an ansible execution method
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/common.yaml]: thanks, fixed.

File: tmt/schemas/discover/shell.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          type: string
+
+        test:
+          type: string
+
+        path:
+          type: string
+
+        duration:
+          $ref: "/schemas/common#/definitions/duration"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/shell.yaml:43]: I will try, the problem is the `name` which only appears here.

File: tmt/schemas/provision/hardware.yml (hunk 4 above)
Comment (happz) [reply tmt/schemas/provision/hardware.yml:218]: With this patch to hardware schema, things start to work:

```patch
diff --git a/tmt/schemas/provision/artemis.yaml b/tmt/schemas/provision/artemis.yaml
index b7c8397..6fabcbb 100644
--- a/tmt/schemas/provision/artemis.yaml
+++ b/tmt/schemas/provision/artemis.yaml
@@ -51,7 +51,7 @@ properties:
     type: string
 
   hardware:
-    $ref: "/schemas/provision/hardware#"
+    $ref: "/schemas/provision/hardware#/definitions/hardware"
 
   pool:
     type: string
diff --git a/tmt/schemas/provision/hardware.yaml b/tmt/schemas/provision/hardware.yaml
index be79a7a..b4a58f4 100644
--- a/tmt/schemas/provision/hardware.yaml
+++ b/tmt/schemas/provision/hardware.yaml
@@ -203,18 +203,13 @@ definitions:
     required:
       - "or"
 
-type: object
-additionalProperties: false
-
-properties:
-  anyOf:
-    - "$ref": "#/definitions/hw_boot"
-    - "$ref": "#/definitions/hw_cpu"
-    - "$ref": "#/definitions/hw_disks"
-    - "$ref": "#/definitions/hw_networks"
-    - "$ref": "#/definitions/hw_virtualization"
-    - "$ref": "#/definitions/hw_block"
-    - "$ref": "#/definitions/hw_and"
-    - "$ref": "#/definitions/hw_or"
-
-  required: []
+  hardware:
+    anyOf:
+      - "$ref": "#/definitions/hw_boot"
+      - "$ref": "#/definitions/hw_cpu"
+      - "$ref": "#/definitions/hw_disks"
+      - "$ref": "#/definitions/hw_networks"
+      - "$ref": "#/definitions/hw_virtualization"
+      - "$ref": "#/definitions/hw_block"
+      - "$ref": "#/definitions/hw_and"
+      - "$ref": "#/definitions/hw_or"
```

1. add a definition of the whole `hardware` key as a definition, not an object described by the schema in `hardware.yaml`
2. this definition is not yet another object but uses `anyOf` directly, it "becomes" one of the listed schemas as needed.

The schema in this PR was probably introducing one extra level of objects by putting this `anyOf` under `properties` in `hardware.yaml`, and this was the issue.

File: tests/unit/test_schemas.py (hunk 7 above)
Comment (happz) [reply tests/unit/test_schemas.py:49]: (FYI, https://github.com/teemtee/tmt/pull/1266 should take care of the "we forgot to list a file" issue)

File: tmt/schemas/discover/shell.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/shell.yaml:43]: I linked all attributes, there is no better way

File: tmt/schemas/plan.yaml [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    $ref: "/schemas/common#/definitions/array_of_strings"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#execute
+  execute:
+    oneOf:
+      - $ref: "/schemas/execute/tmt#"
+      - $ref: "/schemas/execute/upgrade#"
+      - type: array
+        items:
+          $ref: "/schemas/execute/tmt#"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml:99]: ack. added

File: tmt/schemas/plan.yaml [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#finish
+  finish:
+    oneOf:
+      - $ref: "/schemas/finish/ansible#"
+      - $ref: "/schemas/finish/shell#"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml:108]: thanks, added

File: tmt/schemas/plan.yaml [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#link
+  link:
+    $ref: "/schemas/core#/definitions/link"
+
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#order
+  order:
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#finish
+  # TODO: can be a list
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml:124]: indeed, removed

File: tmt/schemas/plan.yaml [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#link
+  link:
+    $ref: "/schemas/core#/definitions/link"
+
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#order
+  order:
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#finish
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml]: thanks, fixed

File: tmt/schemas/plan.yaml [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+      - $ref: "/schemas/provision/connect#"
+      - $ref: "/schemas/provision/container#"
+      - $ref: "/schemas/provision/local#"
+      - $ref: "/schemas/provision/virtual#"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#provision
+  report:
+    oneOf:
+      - $ref: "/schemas/report/display#"
+      - $ref: "/schemas/report/html#"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml:150]: thanks, added

File: tmt/schemas/provision/connect.yaml [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+  password:
+    type: string
+
+  key:
+    type: string
+
+  port:
+    type: integer
+    multipleOf: 1
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/provision/connect.yaml:39]: removed all occurances

File: tmt/schemas/plan.yaml [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+      - $ref: "/schemas/prepare/shell#"
+      - type: array
+        items:
+          anyOf:
+            - $ref: "/schemas/prepare/ansible#"
+            - $ref: "/schemas/prepare/install#"
+            - $ref: "/schemas/prepare/shell#"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#provision
+  # TODO: can be a list
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml]: added array variant, removed TODO

File: tmt/schemas/provision/artemis.yaml [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    type: string
+    enum:
+      - artemis
+
+  name:
+    type: string
+
+  api-retries:
+    type: integer
+    multipleOf: 1
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/provision/artemis.yaml]: Removed, all occurrences, not sure why I thought it makes any sense

File: tmt/schemas/provision/container.yaml [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  how:
+    type: string
+    enum:
+      - container
+
+  name:
+    type: string
+
+  image:
+    type: string
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/provision/container.yaml:26]: Added

File: tmt/schemas/provision/hardware.yml [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+# https://tmt.readthedocs.io/en/stable/spec/plans.html#hardware
+#
+
+$id: /schemas/provision/hardware
+$schema: http://json-schema.org/draft-07/schema
+
+definitions:
+
+  # HW requirements: `boot` block
+  hw_boot:
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/provision/hardware.yml]: removed prefix, I agree it looks better without it. It was kinda copy paste error, I stole this from Artemis.

File: tmt/schemas/provision/hardware.yml [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+      - type: string
+      - type: integer
+
+  # HW requirements: single `network` item
+  hw_network:
+    type: object
+
+    properties:
+      type:
+        type: string
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/provision/hardware.yml:100]: Added them, I would add them now. They are documented.

File: tmt/schemas/provision/hardware.yml (hunk 4 above)
Comment (thrix) [reply tmt/schemas/provision/hardware.yml]: @happz thanks for the investigation and solution, I applied it.

File: tmt/schemas/plan.yaml [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+  context:
+    type: object
+    additionalProperties: false
+
+    properties:
+      arch:
+        $ref: "/schemas/common#/definitions/arch"
+
+      component:
+        type: string
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/schemas/plan.yaml:32]: Yes, fmf is treating all dimensions same way so the value can be one or more strings.

File: tests/unit/test_schemas.py (hunk 7 above)
Comment (thrix) [reply tests/unit/test_schemas.py:49]: @psss thanks, added and renamed

File: tmt/schemas/provision/virtual.yaml (hunk 5 above)
Comment (thrix) [reply tmt/schemas/provision/virtual.yaml:45]: @psss added, please check

File: tmt/schemas/report/html.yaml (hunk 6 above)
Comment (thrix) [reply tmt/schemas/report/html.yaml:20]: added as boolean

File: tmt/schemas/plan.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml:32]: yes, @lukaszachy confirmed on IRC ...

[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/plan.yaml]: it should be fixed

**Comment by thrix** [comment]
> Thanks much for preparing this. This must have been a lot of work! Looks good in general. Added a bunch of comments and questions.

Thanks for the great review and good eyes, I knew I had to miss some things. Should have looked more on module code, I actually used the documentation, which is not actually mentioning some attributes. I believe it is ready for the next round.

CC @lukaszachy @happz

File: tmt/schemas/common.yaml [hunk 23]
```diff
@@ -23,6 +71,17 @@ definitions:
       nick:
         type: string
 
+  # https://tmt.readthedocs.io/en/stable/spec/tests.html#duration
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#shell
+  duration:
+    type: string
+    pattern: "^[0-9]+[smhd]$"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/common.yaml]: made optional

File: tmt/schemas/discover/fmf.yaml [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+    $ref: "/schemas/common#/definitions/fmf_id/properties/ref"
+
+  path:
+    $ref: "/schemas/common#/definitions/fmf_id/properties/path"
+
+  test:
+    $ref: "/schemas/common#/definitions/one_or_more_strings"
+
+  link:
+    $ref: "/schemas/common#/definitions/array_of_strings"
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/fmf.yaml:38]: Filed an issue:| https://github.com/teemtee/tmt/issues/1374

File: tmt/schemas/discover/fmf.yaml [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+  modified-only:
+    type: boolean
+
+  modified-url:
+    type: string
+    format: uri
+
+  modified-ref:
+    type: string
+
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/fmf.yaml:80]: Thanks! added all

File: tmt/schemas/discover/shell.yaml [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+          $ref: "/schemas/test#/properties/test"
+
+        # https://tmt.readthedocs.io/en/stable/spec/tests.html#tier
+        tier:
+          $ref: "/schemas/test#/properties/tier"
+
+      patternProperties:
+        ^extra-:
+          type: string
+
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/shell.yaml:123]: thanks, added!

File: tmt/schemas/discover/shell.yaml [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+  tests:
+    type: array
+    items:
+      type: object
+
+      additionalProperties: false
+
+      properties:
+
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/discover/shell.yaml:39]: it is not possible, we had the same problem including core attributes

File: tmt/schemas/execute/upgrade.yaml [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+additionalProperties: false
+
+properties:
+
+  how:
+    type: string
+    enum:
+      - upgrade
+
+  filter:
```
[the remark this answers is not in the document]
Comment (thrix) [reply tmt/schemas/execute/upgrade.yaml:22]: Added all

File: tmt/schemas/plan.yaml [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+      - type: array
+        items:
+          anyOf:
+            - $ref: "/schemas/provision/artemis#"
+            - $ref: "/schemas/provision/connect#"
+            - $ref: "/schemas/provision/container#"
+            - $ref: "/schemas/provision/local#"
+            - $ref: "/schemas/provision/virtual#"
+
+  # https://tmt.readthedocs.io/en/stable/spec/plans.html#provision
```
Comment (psss) [inline tmt/schemas/plan.yaml:183]:
```suggestion
  # https://tmt.readthedocs.io/en/stable/spec/plans.html#report
```

File: tmt/schemas/plan.yaml [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+  report:
+    oneOf:
+      - $ref: "/schemas/report/display#"
+      - $ref: "/schemas/report/html#"
```
Comment (psss) [inline tmt/schemas/plan.yaml:187]: We're missing the `junit` report plugin.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for addressing all comments. Added two more.

File: tmt/schemas/plan.yaml (hunk 29 above)
Comment (thrix) [reply tmt/schemas/plan.yaml]: fixed

File: tmt/schemas/plan.yaml (hunk 30 above)
Comment (thrix) [reply tmt/schemas/plan.yaml:187]: added junit plugin

**Comment by thrix** [comment]
@lukaszachy @happz @psss addressed all concerns! Thanks for all the comments, feel free to take a look again, happy this finally gets in and unblocks base.py typing

**Review by psss — APPROVED** [review]
Thanks for addressing all comments! Looks good.

**Review by happz — APPROVED** [review]

**Comment by thrix** [comment]
oh my, forgot to squash

**Pull request merged**